### PR TITLE
[v5.x] Cross-compilation to 2.11, 2.12, 2.13

### LINKED ISSUE #806: Cross-compilation to Scala 2.13
We need to cross-compile to both Scala 2.11 and Scala 2.13 (and optionally 2.12). Scala 2.11 is necessary for Android and RoboVM (the later for iOS). 
Sigma unfortunately have diverse dependencies which need to be updated.

Due to Scala 2.11 requirements, there is no simple recipe. Updating to the latest version is often not possible. 
This means that before starting migration, a migration plan should be prepared with the analysis of the situation for each individual dependency.
The migration should be done in v5.0.1 branches (i.e. it will be post v5.0 releases).

### PR DESCRIPTION
**Pull request by aslesarenko**
Closes #806 

This PR makes code changes necessary for cross-compilation to Scala 2.11, 2.12, 2.13

### CODE REVIEW

File: sigmastate/src/main/scala/sigmastate/basics/BcDlogGroup.scala [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 import org.bouncycastle.math.ec.custom.sec.SecP256K1Point
 import org.bouncycastle.math.ec.ECPoint
 import org.bouncycastle.util.BigIntegers
-import spire.syntax.all.cfor
-import scala.collection.mutable
+import debox.cfor
+
+import scala.collection.{Seq, mutable}
 import scala.util.Try
+import scala.collection.compat.immutable.ArraySeq
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/basics/BcDlogGroup.scala:13]: unused import

File: build.sbt [hunk 2]
```diff
@@ -61,20 +63,22 @@ dynverSeparator in ThisBuild := "-"
 val bouncycastleBcprov = "org.bouncycastle" % "bcprov-jdk15on" % "1.64"
 val scrypto            = "org.scorexfoundation" %% "scrypto" % "2.1.10"
 val scorexUtil         = "org.scorexfoundation" %% "scorex-util" % "0.1.8"
-val debox              = "org.spire-math" %% "debox" % "0.8.0"
-val kiama              = "org.bitbucket.inkytonik.kiama" %% "kiama" % "2.1.0"
-val fastparse          = "com.lihaoyi" %% "fastparse" % "1.0.0"
+val debox              = "org.scorexfoundation" %% "debox" % "0.9.0"
+val spireMacros        = "org.typelevel" %% "spire-macros" % "0.17.0-M1"
```
Comment (kushti) [inline build.sbt:67]: Is M1 version really needed? In general, RCx and Mx dependencies are not recommended

File: build.sbt [hunk 3]
```diff
@@ -61,20 +63,22 @@ dynverSeparator in ThisBuild := "-"
 val bouncycastleBcprov = "org.bouncycastle" % "bcprov-jdk15on" % "1.64"
 val scrypto            = "org.scorexfoundation" %% "scrypto" % "2.1.10"
 val scorexUtil         = "org.scorexfoundation" %% "scorex-util" % "0.1.8"
-val debox              = "org.spire-math" %% "debox" % "0.8.0"
-val kiama              = "org.bitbucket.inkytonik.kiama" %% "kiama" % "2.1.0"
-val fastparse          = "com.lihaoyi" %% "fastparse" % "1.0.0"
+val debox              = "org.scorexfoundation" %% "debox" % "0.9.0"
```
Comment (kushti) [inline build.sbt:66]: Not reviewed internal (also, not describing changes from parent library etc)

File: sigmastate/src/main/scala/sigmastate/eval/Extensions.scala [hunk 4]
```diff
@@ -102,5 +103,14 @@ object Extensions {
     def showToString: String = showECPoint(source.asInstanceOf[CGroupElement].wrappedValue)
   }
 
-
+  implicit class DBufferOps[A](val buf: DBuffer[A]) extends AnyVal {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/eval/Extensions.scala:106]: ScalaDoc missed

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Basic.scala [hunk 5]
```diff
@@ -1,42 +1,40 @@
 package sigmastate.lang.syntax
 
-import fastparse.all._
+import fastparse._; import NoWhitespace._
 import fastparse.CharPredicates._
 import scalan.Nullable
 import sigmastate.lang.SourceContext
 import sigmastate.lang.exceptions.SigmaException
 
 object Basic {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Basic.scala:9]: Seems like proper time to add ScalaDoc to Basic

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala [hunk 6]
```diff
@@ -6,21 +6,8 @@ import sigmastate.Values._
 import sigmastate.lang.syntax
 
 trait Core extends syntax.Literals {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:8]: Good time to augment the trait with a ScalaDoc

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 7]
```diff
@@ -1,23 +1,23 @@
...
 import sigmastate.lang._
 import sigmastate.lang.SigmaPredef._
 import sigmastate.lang.syntax.Basic._
 
 import scala.annotation.tailrec
-import scala.collection.mutable
+import scala.collection.compat.immutable.ArraySeq
 
 //noinspection ForwardReference,TypeAnnotation
 trait Exprs extends Core with Types {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:16]: ScalaDoc missed

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 8]
```diff
@@ -30,74 +30,76 @@ trait Exprs extends Core with Types {
...
+        def Else = P( Semi.? ~ `else` ~/ Expr )
         P( Index ~ `if` ~/ "(" ~ ExprCtx.Expr ~ ")" ~ Expr ~ Else ).map {
           case (i, c, t, e) => atSrcPos(i) { mkIf(c.asValue[SBoolean.type], t, e) }
         }
       }
-      val Fun = P(`def` ~ FunDef)
+      def Fun = P(`def` ~ FunDef)
 
-      val LambdaRhs = if (semiInference) P( (Index ~ BlockChunk).map {
+      def LambdaRhs = if (semiInference) P( (Index ~ BlockChunk).map {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:53]: ScalaDoc missed here and everywhere in this trait

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala [hunk 9]
```diff
@@ -14,43 +15,49 @@ object Identifiers {
...
 
-  def VarId0(dollar: Boolean): Parser[Unit] = P( !Keywords ~ Lower ~ IdRest(dollar) )
-  val PlainId: Parser[Unit] = P( !Keywords ~ Upper ~ IdRest(true) | VarId | Operator ~ (!OpChar | &("/*" | "//")) )
-  val PlainIdNoDollar: Parser[Unit] = P( !Keywords ~ Upper ~ IdRest(false) | VarId0(false) | Operator )
-  val BacktickId: Parser[Unit] = P( "`" ~ CharsWhile(NotBackTick) ~ "`" )
-  val Id: P0 = P( BacktickId | PlainId )
+  def VarId0[_:P](dollar: Boolean): P[Unit] = P( !Keywords ~ Lower ~ IdRest(dollar) )
 
-  def IdRest(allowDollar: Boolean): Parser[Unit] = {
+  def UppercaseId[_: P](dollar: Boolean) = P( !Keywords ~ Upper ~ IdRest(dollar) )
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala:26]: ScalaDocs missed in the trait

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Literals.scala [hunk 10]
```diff
@@ -15,81 +14,87 @@ trait Literals { l =>
   import builder._
   def atSrcPos[A](parserIndex: Int)(thunk: => A): A
   def srcCtx(parserIndex: Int): SourceContext
-  def Block: P[Value[SType]]
-  def Pattern: P0
+  
+  def Block[_:P]: P[Value[SType]]
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Literals.scala:18]: Scaladocs missed in the trait

File: sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala [hunk 11]
```diff
@@ -25,33 +23,34 @@ object SigmaParser extends Exprs with Types with Core {
...
-    val TmplStat = P( Prelude ~ BlockDef | StatCtx.Expr )
-    P( "{" ~/ BlockLambda.? ~ Semis.? ~ TmplStat.repX(sep = Semis) ~ Semis.? ~ `}` )
+  def TmplBodyPrelude[_:P] = P( (Annot ~ OneNLMax).rep )
+  def TmplBodyStat[_:P] = P( TmplBodyPrelude ~ BlockDef | StatCtx.Expr )
+
+  def TmplBody[_:P] = {
+    P( "{" ~/ BlockLambda.? ~ Semis.? ~ TmplBodyStat.repX(sep = Semis) ~ Semis.? ~ `}` )
   }
 
 //  val FunDef = {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala:33]: Commented out code

File: sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala [hunk 12]
```diff
@@ -25,33 +23,34 @@ object SigmaParser extends Exprs with Types with Core {
   override def srcCtx(parserIndex: Int): SourceContext =
     SourceContext.fromParserIndex(parserIndex, currentInput.value)
 
-  val TmplBody = {
-    val Prelude = P( (Annot ~ OneNLMax).rep )
-    val TmplStat = P( Prelude ~ BlockDef | StatCtx.Expr )
-    P( "{" ~/ BlockLambda.? ~ Semis.? ~ TmplStat.repX(sep = Semis) ~ Semis.? ~ `}` )
+  def TmplBodyPrelude[_:P] = P( (Annot ~ OneNLMax).rep )
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala:26]: ScalaDocs missed in the trait

File: sigmastate/src/main/scala/sigmastate/lang/Types.scala [hunk 13]
```diff
@@ -10,12 +11,11 @@ import syntax.Basic.error
 
 //noinspection ForwardReference
 trait Types extends Core {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/Types.scala:13]: ScalaDocs missed in the trait

File: sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala [hunk 14]
```diff
@@ -235,7 +241,10 @@ object DataJsonEncoder {
     tpe match {
       case tup: STuple =>
         val tArr = tup.items.toArray
-        val collSource = mutable.ArrayBuilder.make[T#WrappedType]()(tItem.classTag)
+        val collSource = { // this code works both for Scala 2.12 and 2.13
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/serialization/DataJsonEncoder.scala:244]: collSource never used

**Review by kushti — CHANGES_REQUESTED** [review]
You have to get debox dependency reviewed first. And to be reviewed, it should contain description (why forked and what changed).

File: build.sbt (hunk 3 above)
Comment (aslesarenko) [reply build.sbt:66]: the original debox project was abandoned. The PRs there are not reviewed for years.

File: sigmastate/src/main/scala/sigmastate/basics/BcDlogGroup.scala (hunk 1 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/basics/BcDlogGroup.scala:13]: removed

File: build.sbt (hunk 2 above)
Comment (aslesarenko) [reply build.sbt]: clarified

File: sigmastate/src/main/scala/sigmastate/eval/Extensions.scala (hunk 4 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/eval/Extensions.scala:108]: added ScalaDoc for the methods.

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Basic.scala (hunk 5 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Basic.scala:9]: this files have been moved on restructuring PR. Added docs on top of the latest PR here https://github.com/ScorexFoundation/sigmastate-interpreter/commit/9270d42c36b83c53b54748caffc9ec09a742acb5

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala (hunk 9 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala:26]: added ScalaDocs and cleaned the code here https://github.com/ScorexFoundation/sigmastate-interpreter/commit/f6abcaef0f61c5066112eabcb27fb0c2ff5e3243

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Literals.scala (hunk 10 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Literals.scala:18]: done in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/6d2feabfc29a54eb108b77dd09ecf51f51fdf3f1

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala (hunk 6 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:8]: Added docs on top of the latest PR here https://github.com/ScorexFoundation/sigmastate-interpreter/commit/1b09c0fad991b2935742e3a68ae7e890e92d79dd

File: sigmastate/src/main/scala/sigmastate/lang/Types.scala (hunk 13 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/Types.scala:13]: fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/7bfd9e2bc0efb9ab819924d1138d779a2802e86d

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 7 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:16]: fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/2cee01a333ebfa83b24d849b223d5eb89db8bb52

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 8 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:53]: fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/2cee01a333ebfa83b24d849b223d5eb89db8bb52

File: sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala (hunk 12 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala:26]: fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/43101550b16d94172e92a8fcea458a6cfdbe6557

File: sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala (hunk 11 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/SigmaParser.scala:33]: fixed in https://github.com/ScorexFoundation/sigmastate-interpreter/commit/43101550b16d94172e92a8fcea458a6cfdbe6557

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala [hunk 15]
```diff
@@ -29,84 +16,85 @@ trait Core extends syntax.Literals {
...
+//  val `finally`[_:P] = W("finally")
+//  val `do`[_:P] = W("do")
+//  val `yield`[_:P] = W("yield")
+//  val `while`[_:P] = W("while")
+//  val `<%`[_:P] = O("<%")
+//  val `override`[_:P] = W("override")
+//  val `forSome`[_:P] = W("forSome")
+//  val `for`[_:P] = W("for")
+//  val `abstract`[_:P] = W("abstract")
+//  val `throw`[_:P] = W("throw")
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:62]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala [hunk 16]
```diff
@@ -29,84 +16,85 @@ trait Core extends syntax.Literals {
...
+  def `lazy`[_:P] = W("lazy")
+  def `>:`[_:P] = O(">:")
+  def `<:`[_:P] = O("<:")
 //  val `final` =  W("final")
-//  val `sealed` = W("sealed")
-//  val `private` = W("private")
-//  val `protected` = W("protected")
+//  val `sealed`[_:P] = W("sealed")
+//  val `private`[_:P] = W("private")
+//  val `protected`[_:P] = W("protected")
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:69]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala [hunk 17]
```diff
@@ -29,84 +16,85 @@ trait Core extends syntax.Literals {
...
-  val StableId = {
-//    val ClassQualifier = P( "[" ~ Id ~ "]" )
-//    val ThisSuper = P( `this` | `super` ~ ClassQualifier.? )
-//    val ThisPath: P0 = P( ThisSuper ~ ("." ~ PostDotCheck ~/ Id).rep )
-    val IdPath = P( Index ~ Id.! ~ ("." ~ PostDotCheck ~/ Index ~ (`this`.! | Id.!)).rep /*~ ("." ~ ThisPath).?*/ ).map {
+  def PostDotCheck[_:P]: P0 = P( WL ~ !(`super` | `this` | "{" |  `_` | `type`) )
+  def StableId[_:P] = {
+//    val ClassQualifier[_:P] = P( "[" ~ Id ~ "]" )
+//    val ThisSuper[_:P] = P( `this` | `super` ~ ClassQualifier.? )
+//    val ThisPath: P0[_:P] = P( ThisSuper ~ ("." ~ PostDotCheck ~/ Id).rep )
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:96]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala [hunk 18]
```diff
@@ -29,84 +16,85 @@ trait Core extends syntax.Literals {
...
+    def IdPath = P( Index ~ Id.! ~ ("." ~ PostDotCheck ~/ Index ~ (`this`.! | Id.!)).rep /*~ ("." ~ ThisPath).?*/ ).map {
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:97]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 19]
```diff
@@ -30,74 +30,76 @@ trait Exprs extends Core with Types {
   // Expressions directly within a `val x = ...` or `def x = ...`
   object FreeCtx extends WsCtx(semiInference=true, arrowTypeAscriptions=true)
 
-  val TypeExpr = ExprCtx.Expr
+  def TypeExpr[_:P] = ExprCtx.Expr
 
   private val predefFuncRegistry = new PredefinedFuncRegistry(builder)
   import predefFuncRegistry._
 
   //noinspection TypeAnnotation,ForwardReference
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:38]: ???

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 20]
```diff
@@ -30,74 +30,76 @@ trait Exprs extends Core with Types {
...
 
-      val LambdaRhs = if (semiInference) P( (Index ~ BlockChunk).map {
+      def LambdaRhs = if (semiInference) P( (Index ~ BlockChunk).map {
         case (index, (_ , b))  => atSrcPos(index) { block(b) }
       } )
       else P( Expr )
 //      val ParenedLambda = P( Parened ~~ (WL ~ `=>` ~ LambdaRhs.? /*| ExprSuffix ~~ PostfixSuffix ~ SuperPostfixSuffix*/) ).map {
 //        case (args, None) => mkLambda(args, UnitConstant)
 //        case (args, Some(body)) => mkLambda(args, body)
 //      }
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:60]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 21]
```diff
@@ -30,74 +30,76 @@ trait Exprs extends Core with Types {
...
-      val PostfixLambda = P( Index ~ PostfixExpr ~ (`=>` ~ LambdaRhs.? | SuperPostfixSuffix).? ).map {
+      def PostfixLambda = P( Index ~ PostfixExpr ~ (`=>` ~ LambdaRhs.? | SuperPostfixSuffix).? ).map {
         case (_, e, None) => e
         case (_, e, Some(None)) => e
         case (i, Tuple(args), Some(Some(body))) => atSrcPos(i) { lambda(args, body) }
         case (i, e, Some(body)) => error(s"Invalid declaration of lambda $e => $body", Some(srcCtx(i)))
       }
-      val SmallerExprOrLambda = P( /*ParenedLambda |*/ PostfixLambda )
+      def SmallerExprOrLambda = P( /*ParenedLambda |*/ PostfixLambda )
 //      val Arg = (Id.! ~ `:` ~/ Type).map { case (n, t) => mkIdent(IndexedSeq(n), t)}
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:68]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala [hunk 22]
```diff
@@ -109,8 +111,8 @@ trait Exprs extends Core with Types {
         }
     }
 
-    val Parened = P ( "(" ~/ TypeExpr.repTC() ~ ")" )
-    val SimpleExpr = {
+    def Parened[_:P] = P ( "(" ~/ TypeExpr.rep(0, ",") ~ TrailingComma ~ ")" )
+    def SimpleExpr[_:P] = {
 //      val New = P( `new` ~/ AnonTmpl )
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:116]: commented out code

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala [hunk 23]
```diff
@@ -14,43 +15,49 @@ object Identifiers {
...
 
-  val alphaKeywords = Seq(
+  final val alphaKeywords = Seq(
     "case", "else", "false", "function", "if", "match", "return", "then", "true"
   )
-  val AlphabetKeywords: Parser[Unit] = P {
-    StringIn(alphaKeywords:_*) ~ !Letter
+  def AlphabetKeywords[_:P]: P[Unit] = P {
+    StringIn("case", "else", "false", "function", "if", "match", "return", "then", "true") ~
+//    ("case" | "else" | "false" | "function" | "if" | "match" | "return" | "then" | "true") ~
```
Comment (kushti) [inline sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala:49]: commented out code

**Review by kushti — CHANGES_REQUESTED** [review]

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala (hunk 15 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:62]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala (hunk 16 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:69]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala (hunk 17 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:96]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala (hunk 18 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Core.scala:97]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 19 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:38]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 20 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:60]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 21 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:68]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala (hunk 22 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Exprs.scala:116]: Fixed in #853

File: sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala (hunk 23 above)
Comment (aslesarenko) [reply sigmastate/src/main/scala/sigmastate/lang/syntax/Identifiers.scala:49]: Fixed in #853

**Review by kushti — CHANGES_REQUESTED** [review]
I thinking it is good time to add building instructions to readme as well (including how to build for all the supported Scala versions)

File: common/src/main/scala/scalan/util/CollectionUtil.scala [hunk 24]
```diff
@@ -222,8 +222,8 @@ object CollectionUtil {
       res
     }
 
-    def filterCast[B:ClassTag](implicit cbf: CanBuildFrom[Source[A], B, Source[B]]): Source[B] = {
-      val b = cbf()
+    def filterCast[B:ClassTag](implicit cbf: BuildFrom[Source[A], B, Source[B]]): Source[B] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/src/main/scala/scalan/util/CollectionUtil.scala:225]: method doesn't exist in latest version (removed in subsequent PRs)

File: common/src/main/scala/scalan/util/CollectionUtil.scala [hunk 25]
```diff
@@ -278,23 +278,23 @@ object CollectionUtil {
     }
 
     def mapUnzip[B1, B2, B3](f: A => (B1,B2,B3))
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/src/main/scala/scalan/util/CollectionUtil.scala:280]: method doesn't exist in latest version (removed in subsequent PRs)

File: common/src/main/scala/scalan/util/CollectionUtil.scala [hunk 26]
```diff
@@ -278,23 +278,23 @@ object CollectionUtil {
...
+      val b3 = cbf3.newBuilder(xs)
       for (x <- xs) {
         val (y1, y2, y3) = f(x)
         b1 += y1; b2 += y2; b3 += y3
       }
       (b1.result(), b2.result(), b3.result())
     }
 
-    def distinctBy[K](key: A => K)(implicit cbf: CanBuildFrom[Source[A], A, Source[A]]): Source[A] = {
+    def distinctBy[K](key: A => K)(implicit cbf: BuildFrom[Source[A], A, Source[A]]): Source[A] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/src/main/scala/scalan/util/CollectionUtil.scala:295]: method already documented latest version (docs added  in subsequent PRs)

File: common/src/main/scala/scalan/util/CollectionUtil.scala [hunk 27]
```diff
@@ -308,22 +308,22 @@ object CollectionUtil {
...
         result.update(key, if (result.contains(key)) reduce(result(key), value) else value)
         result
       })
-      val b = cbf()
+      val b = cbf.newBuilder(xs)
       for (kv <- result) b += kv
       b.result()
     }
 
     def mergeWith[K]
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/src/main/scala/scalan/util/CollectionUtil.scala:323]: method doesn't exist in latest version (removed in subsequent PRs)

File: common/src/main/scala/scalan/util/CollectionUtil.scala [hunk 28]
```diff
@@ -234,7 +234,7 @@ object CollectionUtil {
       b.result()
     }
 
-    def cast[B:ClassTag](implicit cbf: CanBuildFrom[Source[A], B, Source[B]]): Source[B] = {
+    def cast[B:ClassTag](implicit cbf: BuildFrom[Source[A], B, Source[B]]): Source[B] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply common/src/main/scala/scalan/util/CollectionUtil.scala:237]: Added ScalaDoc to base classes in #853

File: library/src/main/scala/wrappers/scalan/impl/WRTypesImpl.scala [hunk 29]
```diff
@@ -35,7 +35,7 @@ object WRType extends EntityObject("WRType") {
     override def name: Ref[String] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply library/src/main/scala/wrappers/scalan/impl/WRTypesImpl.scala:35]: this code is part of graph-IR machinery. It is better to document it in a separate PR on top of the latest version (there is an issue #845). Also it is not in the scope of this PR.

File: sigma-library/src/main/scala/special/sigma/impl/SigmaDslImpl.scala [hunk 30]
```diff
@@ -47,49 +47,49 @@ object BigInt extends EntityObject("BigInt") {
     override def toByte: Ref[Byte] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigma-library/src/main/scala/special/sigma/impl/SigmaDslImpl.scala:47]: Same as above

File: sigmastate/src/main/scala/org/ergoplatform/dsl/ContractSyntax.scala [hunk 31]
```diff
@@ -25,11 +25,11 @@ trait ContractSyntax { contract: SigmaContract =>
                   dslSpec: Proposition,
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/dsl/ContractSyntax.scala:25]: documented ContractSyntax

File: sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala [hunk 32]
```diff
@@ -98,7 +98,7 @@ object ValidationRules {
       if (!root.tpe.isSigmaProp) {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply sigmastate/src/main/scala/org/ergoplatform/validation/ValidationRules.scala:98]: each rule has description and also error message. 
Also added ScalaDoc for the apply method in #853

File: library/src/main/scala/wrappers/scala/impl/WOptionsImpl.scala [hunk 33]
```diff
@@ -41,14 +41,14 @@ object WOption extends EntityObject("WOption") {
     override def isEmpty: Ref[Boolean] = {
```
[the remark this answers is not in the document]
Comment (aslesarenko) [reply library/src/main/scala/wrappers/scala/impl/WOptionsImpl.scala:41]: this code is part of graph-IR machinery. It is better to document it in a separate PR on top of the latest version (there is an issue #845). Also it is not in the scope of this PR.

**Review by kushti — DISMISSED** [review]

**Pull request merged**